Order messages thread link was modified to use symfony controller route

tdavidsonas88 opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Order messages customer thread link was modified to use symfony controller route
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10556
| How to test?  | Rebuilding assets is required. From the Front Office go to Order history -> Details and then select product, enter message and send it. Login to the Back Office -> Notification dropdown toggle (bell icon top right corner) -> Messages -> should be able to go to the migrated controller link and see customer thread messages. If you do not see the new links, please go to Advanced Parameters -> Performance -> Clear Cache

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/16678)
<!-- Reviewable:end -->

tdavidsonas88 committed:
Order messages thread link was modified to go to symfony controller

matks commented on admin-dev/themes/new-theme/js/notifications.js:
```diff
@@ -63,26 +65,31 @@ const refreshNotifications = function () {
+        .replace(/_date_add_/g, value.date_add)
+        .replace(/_status_/g, value.status)
+        .replace(/order_url/g, `${baseAdminDir}index.php?tab=AdminOrders&token=${token_admin_orders}&vieworder&id_order=${value.id_order}`)
+        .replace(/customer_url/g, `${baseAdminDir}index.php?tab=AdminCustomers&token=${token_admin_customers}&viewcustomer&id_customer=${value.id_customer}`);
```
Nice catch @tdavidsonas88 ! I see there is also a link using legacy customer route although Customers page has been migrated as well. Could you please do the same fix for line 85 ? 😉

tdavidsonas88 replied:
Thank you @matks !  committed the change for legacy customer route

tdavidsonas88 committed:
link using legacy customer route replaced with Symfony route; replaceGlobally function created;

PierreRambaud commented on admin-dev/themes/new-theme/js/notifications.js:
```diff
@@ -22,6 +22,8 @@
  * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
  * International Registered Trademark & Property of PrestaShop SA
  */
+import Router from "./components/router";
```
```suggestion
import Router from './components/router';
```

PierreRambaud commented on admin-dev/themes/new-theme/js/notifications.js:
```diff
@@ -60,35 +62,50 @@ const refreshNotifications = function () {
+        .replace(/_status_/g, value.status)
+        .replace(/order_url/g, `${baseAdminDir}index.php?tab=AdminOrders&token=${token_admin_orders}&vieworder&id_order=${value.id_order}`)
+
+      let customerThreadId = parseInt(value.id_customer_thread);
```
```suggestion
      const customerThreadId = parseInt(value.id_customer_thread, 10);
```

PierreRambaud commented on admin-dev/themes/new-theme/js/notifications.js:
```diff
@@ -60,35 +62,50 @@ const refreshNotifications = function () {
+        tplReplaced = replaceGlobally(tplReplaced, 'message_url', route);
+      }
+
+      let customerId = parseInt(value.id_customer);
```
```suggestion
      const customerId = parseInt(value.id_customer, 10);
```

PierreRambaud commented on admin-dev/themes/new-theme/js/notifications.js:
```diff
@@ -60,35 +62,50 @@ const refreshNotifications = function () {
+
+      let customerId = parseInt(value.id_customer);
+      if (customerId) {
+        let route = router.generate('admin_customers_view', {customerId});
```
You should declare `route` and `tplReplaced` once

tdavidsonas88 replied:
declared outside foreach, hope it is what you expected. If not, let me know

PierreRambaud commented on admin-dev/themes/new-theme/js/notifications.js:
```diff
@@ -60,35 +62,50 @@ const refreshNotifications = function () {
+      let tplReplaced = tpl.replace(/_id_order_/g, parseInt(value.id_order))
+        .replace(/_customer_name_/g, value.customer_name)
+        .replace(/_iso_code_/g, value.iso_code)
+        .replace(/_carrier_/g, (value.carrier !== "" ? ` - ${value.carrier}` : ""))
```
```suggestion
        .replace(/_carrier_/g, (value.carrier !== '' ? ` - ${value.carrier}` : ''))
```

PierreRambaud commented on admin-dev/themes/new-theme/js/notifications.js:
```diff
@@ -60,35 +62,50 @@ const refreshNotifications = function () {
+        .replace(/_carrier_/g, (value.carrier !== "" ? ` - ${value.carrier}` : ""))
+        .replace(/_total_paid_/g, value.total_paid)
+        .replace(/_id_customer_/g, parseInt(value.id_customer))
+        .replace(/_company_/g, (value.company !== "" ? ` (${value.company}) ` : ""))
```
```suggestion
        .replace(/_company_/g, (value.company !== '' ? ` (${value.company}) ` : ''))
```

tdavidsonas88 committed:
code refactored after QA; switch created for symfony url generation

PierreRambaud commented on admin-dev/themes/new-theme/js/notifications.js:
```diff
@@ -64,36 +66,55 @@ let fillTpl = function (results, eltAppendTo, tpl) {
   if (results.length > 0) {
     eltAppendTo.removeClass('empty');
+
+    let router = new Router();
```
```suggestion
    const router = new Router();
```

PierreRambaud replied:
Sorry last thing :)

tdavidsonas88 committed:
code refactored after QA; router

Robin-Fischer-PS commented:
Hi @tdavidsonas88 :)

With the PR installed, and the assets rebuild (BTW, don't hesitate to specify that rebuilding assets is required in the "How to test"), the link from the notifications still redirect me to Legacy page and not to migrated page.

Could you take a look on this bug ?

Thanks !

tdavidsonas88 commented:
> Hi @tdavidsonas88 :)
> 
> With the PR installed, and the assets rebuild (BTW, don't hesitate to specify that rebuilding assets is required in the "How to test"), the link from the notifications still redirect me to Legacy page and not to migrated page.
> 
> Could you take a look on this bug ?
> 
> Thanks !

Hi @Robin-Fischer-PS :)

After rebuilding the assets please check if you see in your browser sources: main.bundle.js in admin-dev/themes/new-theme/public . If you don't see it, then go to Advanced Parameters -> Performance -> Clear Cache. If that does not help and the code is merged then please let me know.

Also if the route is not found then run `bin/console fos:js-routing:dump --format=json --target=admin-dev/themes/new-theme/js/fos_js_routes.json`

Robin-Fischer-PS commented:
Hi @tdavidsonas88 ! With the cache cleared it's now OK !
However, as soon as I go back to any Legacy page, the "main.bundle.js" source disappears and the link is wrong again (this happens when Cache is enabled, and also when cache is disabled).

I don't think this is OK, but maybe I'm missing something

tdavidsonas88 commented:
Hi @Robin-Fischer-PS @matks ! Would it be good, if I change default theme links to use these new controllers? (not sure about the backward compatibility) . Default theme has a duplicate file notifications.js which has these legacy links also.

tdavidsonas88 committed:
change default theme legacy links also

tdavidsonas88 commented:
Hi @Robin-Fischer-PS @matks ! changed default theme legacy links also

PierreRambaud commented on js/admin/notifications.js:
```diff
@@ -116,7 +116,7 @@ function getPush()
 				html = "";
 				$.each(json.customer.results, function(property, value) {
-					html += "<a class='notif' href='" + value.customer_view_url + "'>";
+					html += "<a class='notif' href='"+baseAdminDir+"index.php/sell/customers/" + parseInt(value.id_customer, 10) +"/view?_token=" + token_admin_customers + "'>";
```
it's not a good idea to have route in plain text in the javascript. Please use the router component instead

tdavidsonas88 committed:
_legacy_link property enabled in customer_threads.yml

tdavidsonas88 commented on src/Core/Domain/Notification/QueryResult/NotificationResult.php:
```diff
@@ -119,7 +125,8 @@ public function __construct(
         string $status,
-        string $dateAdd
+        string $dateAdd,
+        string $customerThreadViewUrl = ''
```
@matks reviewed the solution with @rokaszygmantas , would be nice if somebody confirm this does not break backward compatibiliy as the new property was added in order to use _legacy_link in customer_threads.yml

rokaszygmantas replied:
we intentionally left this `$customerThreadViewUrl` property optional in constructor, to not break BC, but maybe that's not a problem? @matks

tdavidsonas88 committed:
regenerated the routes

PierreRambaud commented:
Look like something went wrong with your rebase :thinking:

tdavidsonas88 committed:
fixes after rebase from develop

tdavidsonas88 commented:
> Look like something went wrong with your rebase

Yes, fixing I thought this goes to 1.7.7.x and rebased to it. Then reset hard to previous commit and rebased from develop

tdavidsonas88 committed:
deleted debug statements

tdavidsonas88 commented:
> Look like something went wrong with your rebase

@PierreRambaud Rebased correctly now, fixed conflicts, ready for review; No longer blocking this one as a dev :slightly_smiling_face:

tdavidsonas88 committed:
not used router removed after rebase

NeOMakinG commented:
@tdavidsonas88 Hey, can you fix ESLint issues before we review ? Thanks :)

tdavidsonas88 committed:
js fixes

tdavidsonas88 committed:
js fixes

tdavidsonas88 committed:
js fixes for foreach function

tdavidsonas88 committed:
prefer-arrow-callback for linter fixed

tdavidsonas88 committed:
prefer-arrow-callback for linter fixed

tdavidsonas88 committed:
just to rerun failing cs-fixer linter

tdavidsonas88 commented:
No longer blocked by me. Label can be changed! :slightly_smiling_face:

NeOMakinG commented on admin-dev/themes/new-theme/js/notifications.js:
```diff
@@ -22,13 +22,14 @@
     type: 'POST',
     headers: {'cache-control': 'no-cache'},
-    url: `${window.adminNotificationGetLink}&rand=${new Date().getTime()}`,
+    url: `${window.admin_notification_get_link}&rand=${new Date().getTime()}`,
```
Just a question, can't find why, why not keeping camelCase as it's cleaner ?

matks committed:
Update FOS JS routes JSON file

matks commented:
PR rebased (and approvals gone because of that)

matks commented:
I had to rebuild this PR to move its base from `develop` to `1.7.7.x` https://github.com/PrestaShop/PrestaShop/pull/19385/
